### ISSUE TITLE
"forceReprojection" setting not included in default.vrsettings, logs being spammed

### ISSUE THREAD

**Issue by username223** [OpenVR-Advanced-Settings/OpenVR-AdvancedSettings#25]
## Setup
* Versions: 
  * At least git tag "v2.7", currently in master
  * SteamVR:
    ```
    Built dec 4, 2018 at 23:11
    Version 1.1.4 (1543961514)
    ```
* HMD: HTC Vive with Wireless Adapter Kit
* OS: Windows 10
* Severity: Medium, could be indication of faulty feature.
* Reproducible: Independent reproduction [here](https://www.reddit.com/r/SteamVR/comments/adm1ps/openvr_advancedsettings_error/) on v2.7.

## Summary

Either the `C:\Program Files (x86)\Steam\config\steamvr.vrsettings` or the `C:\Program Files (x86)\Steam\steamapps\common\SteamVR\resources\settings\default.vrsettings` file requires the line 
```
"forceReprojection": true,
```
under the `"steamvr": {` block.

It is currently absent from mine and this [reddit posters](https://www.reddit.com/r/SteamVR/comments/adm1ps/openvr_advancedsettings_error/) config.

The line has been documented as having been present in [this](https://github.com/daftpwner/lighthouse_ros/blob/0303e5ec35903525f49b8e05caa2c8c5ad6efd76/default.vrsettings#L35) file from 2017-07-14. It is unknown why and when it was removed, or if different configurations have different default values.

### 1. Programatically change the `steamvr.vrsettings` file

While this would solve the problem in a direct way, parsing proprietary Valve config formats is _way_ outside the scope of this project and changing third party (Valve being considered the third party) config files for all users is definitely not something to do lightly.

### 2. Wait for Valve to fix the settings file

Wait (and contact Valve for a fix).

#### 2.1 Keep the log message

Add a comment to around the log code and wait. Doesn't require much effort, but logs continue to be polluted by useless logs every ~2 seconds.

#### 2.2 Temporarily suppress the log message until a fix has been released

Logic can be added to `SteamVRTabController.cpp` with a conditional to filter out this specific message. Comments should also be added pointing to this issue. This is my recommended solution.

## Details

This message is logged every ~2 seconds for the duration of the application runtime.

```
[WARNING] 2019-01-07 22:05:00: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:02: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:03: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:05: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:07: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:08: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:10: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:11: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:13: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
[WARNING] 2019-01-07 22:05:15: Could not read "forceReprojection" setting: VRSettingsError_UnsetSettingHasNoDefault
```

Searching the code base leads to this on line `229-241` of `SteamVRTabController.cpp`:

```c++
auto fr = vr::VRSettings()->GetBool(
    vr::k_pch_SteamVR_Section,
    vr::k_pch_SteamVR_ForceReprojection_Bool,
    &vrSettingsError );
if ( vrSettingsError != vr::VRSettingsError_None )
{
    LOG( WARNING ) << "Could not read \""
                   << vr::k_pch_SteamVR_ForceReprojection_Bool
                   << "\" setting: "
                   << vr::VRSettings()->GetSettingsErrorNameFromEnum(
                          vrSettingsError );
}
setForceReprojection( fr );

```

Which is almost definitely the source of the error message.

Inspecting the `vr::VRSettings()->GetBool()` function declaration reveals the following in `openvr.h` on line `1938-1940` (emphasis on comments):

```c++
// Users of the system need to provide a proper default in default.vrsettings in the resources/settings/ directory
// of either the runtime or the driver_xxx directory. Otherwise the default will be false, 0, 0.0 or ""
virtual bool GetBool( const char *pchSection, const char *pchSettingsKey, EVRSettingsError *peError = nullptr ) = 0;
```

Inside the `Steam\steamapps\common\SteamVR\` directory there is a `resources\settings` directory. It contains `default.vrsettings` .

Adding the line 
```
"forceReprojection": true,
```
inside the `"steamvr": {` block solves the error. Removing the line again makes the bug appear again.

Supposedly this file contains the defaults. Changing the defaults would not be wise. Additionally, the file at `C:\Program Files (x86)\Steam\config\steamvr.vrsettings` contains approximately the same layout. Adding the same setting under the same header also solves the bug. Removing it reintroduces the bug.

The `steamvr.vrsettings` file appears to be programatically changed when running SteamVR.

After some searching it is found that [this](https://github.com/daftpwner/lighthouse_ros/blob/0303e5ec35903525f49b8e05caa2c8c5ad6efd76/default.vrsettings#L35) file from 2017-07-14 has the line `"forceReprojection" : false,` in it. 

It is unknown exactly when the `"forceReprojection"` line was dropped from the `defaults.vrsettings` file, or if different configurations and hardware setups have different default files.

**Comment by ykeara** [comment]
Valve removed the  keys in openvr 1.1.3
	static const char * const k_pch_SteamVR_AllowAsyncReprojection_Bool = "allowAsyncReprojection";
	static const char * const k_pch_SteamVR_AllowReprojection_Bool = "allowInterleavedReprojection";
	static const char * const k_pch_SteamVR_ForceReprojection_Bool = "forceReprojection";

and Added the keys:

	static const char * const k_pch_SteamVR_MotionSmoothing_Bool = "motionSmoothing";
	static const char * const k_pch_SteamVR_MotionSmoothingOverride_Int32 = "motionSmoothingOverride";

as far as I can tell only control over re projection type settings is motion smoothing on/off, as asynch reprojection appears to be always on, and no way to toggle in steamVR or openVR api.

Will remove feature from ADV settings.

**Comment by ykeara** [comment]
Resolved In PR #30
